pkg/cli/cli.go

```diff
@@ -248,20 +335,90 @@ func (c cli) buildRootCmd() *cobra.Command {
 	return rootCmd
 }
 
+// filterPluginsByKeys matches plugins known to a cli to a set of plugin keys,
+// returning unambiguously matched plugins.
+// A match occurs if:
+// - Name and version are the same.
+// - Long or short name is the same, and only one version for that name exists.
+// If long or short name is the same, but multiple versions for that name exist,
+// don't guess which version to use and instead return an error.
+func filterPluginsByKeys(versionedPlugins []plugin.Base, cliPluginKeys map[string]string) ([]plugin.Base, error) {
+	if len(cliPluginKeys) == 0 {
+		return versionedPlugins, nil
+	}
+
+	// Find all potential mateches for this plugin's name.
+	definitelyMatch, maybeMatch := []plugin.Base{}, map[string][]plugin.Base{}
+	for pluginName, pluginVersion := range cliPluginKeys {
+		cliPluginKey := plugin.Key(pluginName, pluginVersion)
+		// Prevents duplicate versions with the same name from being added to
+		// maybeMatch.
+		hasDefinitely := false
+		for _, versionedPlugin := range versionedPlugins {
+			longName := versionedPlugin.Name()
+			shortName := plugin.GetShortName(longName)
+
+			switch {
+			// Exact match.
+			case pluginName == longName && pluginVersion == versionedPlugin.Version():
+				definitelyMatch = append(definitelyMatch, versionedPlugin)
+				hasDefinitely = true
+			// Is at least a name match, and the CLI version wasn't set or was set
+			// and matches.
+			case pluginName == longName || pluginName == shortName:
+				if !hasDefinitely && (pluginVersion == "" || pluginVersion == versionedPlugin.Version()) {
+					maybeMatch[cliPluginKey] = append(maybeMatch[cliPluginKey], versionedPlugin)
```

Comment:
In case of no match being found, neither a full match nor a partial match, there will be no plugin added to the list and no `maybeMatch` to try to educately add one. Shouldn't this be considered an error? For example, imagine I misspell `gu/v2`. I feel like that should definetely return an error.

It should be as easy as checking if `hasDefinetely` is false and there is no entry in `maybeMatch` when the inner loop ends.

Reply:
An error is returned if a plugin key doesn't match anything.